recipe for Multi-Decoder DPRNN

JunzheJosephZhu opened the pull request:
Don't have the bash script and separate function yet

mpariente commented:
All the things that were already merge shouldn't appear. Please merge or rebase master.

JunzheJosephZhu commented:
I already rebased, I have no clue why it still shows

On Sun, Mar 14, 2021 at 4:27 AM Pariente Manuel ***@***.***>
wrote:

> All the things that were already merge shouldn't appear. Please merge or
> rebase master.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-798780531>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF2C6G6RIYYWG7MDEMPS5Z3TDPDDFANCNFSM4ZCSQUPQ>
> .
>
-- 
Joseph Zhu

mpariente commented:
There might be commits in the middle, IDK
Can you apply the new commits on top of master for example?

Reviewing this PR doubles the effort, sorry but I won't do it until the
diff is normal.

Le sam. 13 mars 2021 à 22:35, Joseph ***@***.***> a écrit :

> I already rebased, I have no clue why it still shows
>
> On Sun, Mar 14, 2021 at 4:27 AM Pariente Manuel ***@***.***>
> wrote:
>
> > All the things that were already merge shouldn't appear. Please merge or
> > rebase master.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/asteroid-team/asteroid/pull/463#issuecomment-798780531
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AF2C6G6RIYYWG7MDEMPS5Z3TDPDDFANCNFSM4ZCSQUPQ
> >
> > .
> >
> --
> Joseph Zhu
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-798789228>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEND2HC3WGYHKLVW5IEGTHTTDPLAVANCNFSM4ZCSQUPQ>
> .
>

JunzheJosephZhu commented:
> All the things that were already merge shouldn't appear. Please merge or rebase master.

Do you mean these commits shouldn't appear? There are only 9 files changed in the "files changed" section, could you not review these?

mpariente commented:
Yes, I can review now because the diff was fixed. 
Make a soft link to the utils folder with `ln -s` please, so that we don't have those duplicates. 
I can review after that.

JunzheJosephZhu commented:
What name should I softlink the utils folder to? Which other file is
referring to the utils folder?

On Tue, Mar 16, 2021 at 12:17 AM Pariente Manuel ***@***.***>
wrote:

> Yes, I can review now because the diff was fixed.
> Make a soft link to the utils folder with ln -s please, so that we don't
> have those duplicates.
> I can review after that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-799549312>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF2C6G57NO6YFXU35C2KG6LTDYXKHANCNFSM4ZCSQUPQ>
> .
>
-- 
Joseph Zhu

mpariente commented:
Like the others recipes, softlink utils to utils. 
Have a look at the other recipes

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/local/conf.yml:
```diff
@@ -0,0 +1,42 @@
+  weight_decay: 0.00000
+# Data config
+data:
+  dataset: "/ws/ifp-10_3/hasegawa/junzhez2/asteroid/dataset"
```
There should be no abs path in config

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/local/conf.yml:
```diff
@@ -0,0 +1,42 @@
+  seglen: 4.0
+  minlen: 2.0
+loss:
+  lambda: 0.05
```
Need new line.

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/run.sh:
```diff
@@ -0,0 +1,137 @@
+	unzip ./local/create-speaker-mixtures.zip -d ./local/create-speaker-mixtures
+	mv ./local/create-speaker-mixtures.zip ./local/create-speaker-mixtures
+
+	echo "You need to generate the wsj0-mix dataset using the official MATLAB
+			  scripts (already downloaded into ./local/create-speaker-mixtures).
+			  If you don't have Matlab, you can use Octavve and replace
+				all mkdir(...) in create_wav_2speakers.m with system(['mkdir -p '...]).
+				Note: for 2-speaker separation, the sep_clean task from WHAM is the same as
+				wsj0-2mix and the mixing scripts are in Python.
+				Specify wsj0mix_wav_dir and start from stage 2 when the mixtures have been generated.
+				Exiting now."
+	exit 1
```
Fix indent

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/run.sh:
```diff
@@ -0,0 +1,137 @@
+if [[ $stage -le  2 ]]; then
+	# Make json directories with min/max modes and sampling rates
+	echo "Stage 2: Generating json files including wav path and duration"
+	for tmp_nsrc in 2 3 4 5; do
+				tmp_dumpdir=$wsj0mix_wav_dir/${tmp_nsrc}speakers/wav8k/min
+				echo "Generating json files in $tmp_dumpdir"
+				[[ ! -d $tmp_dumpdir ]] && mkdir -p $tmp_dumpdir
+				local_wsj_dir=$wsj0mix_wav_dir/${tmp_nsrc}speakers/wav8k/min/
+				$python_path local/preprocess_wsj0mix.py --in_dir $local_wsj_dir \
+				 																			--n_src $tmp_nsrc \
```
Fix indent

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/run.sh:
```diff
@@ -0,0 +1,137 @@
+
+if [[ $stage -le 4 ]]; then
+	echo "Stage 4 : Evaluation"
+	CUDA_VISIBLE_DEVICES=$id $python_path eval.py \
+	  --n_src $n_src \
+		--test_dir $test_dir \
+		--use_gpu $eval_use_gpu \
+		--exp_dir ${expdir} | tee logs/eval_${tag}.log
+	cp logs/eval_${tag}.log $expdir/eval.log
```
Fix indent

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/wsj0_mix_variable.py:
```diff
@@ -0,0 +1,157 @@
+        return mixture, sources
+
+
+def _collate_fn(batch):
+    """
```
More explicit name and a simple docstring saying what it does.

mpariente reviewed (commented):
Mainly looks good to me. Just few fixes and we can merge. 

Thanks !

JunzheJosephZhu commented:
how do you resume training btw, if a session is interrupted? It seems that people are manually typing the checkpoint into the trainer?

mpariente commented:
If the experiment name is the same, PL handles it for you (depending on the version)

mpariente commented:
Let's get this merged? There are very few things to fix :-)

JunzheJosephZhu commented:
Sorry! I was busy with work and forgot. Will do it right away. Btw, I still
need to write the bash scipt/inference part. Do I use another commit?

On Thu, Mar 25, 2021 at 10:17 PM Pariente Manuel ***@***.***>
wrote:

> Let's get this merged? There are very few things to fix :-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-806833417>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF2C6GYF6GA3SWRQ46NZNGLTFNAWNANCNFSM4ZCSQUPQ>
> .
>
-- 
Joseph Zhu

JunzheJosephZhu commented:
It seems that I'm unable to connect to my server with ssh, because my new
apartment has horrible internet connection... Let me try to figure this out.
Joseph Zhu


On Thu, Mar 25, 2021 at 11:56 PM Joseph Zhu ***@***.***>
wrote:

> Sorry! I was busy with work and forgot. Will do it right away. Btw, I
> still need to write the bash scipt/inference part. Do I use another commit?
>
> On Thu, Mar 25, 2021 at 10:17 PM Pariente Manuel ***@***.***>
> wrote:
>
>> Let's get this merged? There are very few things to fix :-)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-806833417>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AF2C6GYF6GA3SWRQ46NZNGLTFNAWNANCNFSM4ZCSQUPQ>
>> .
>>
> --
> Joseph Zhu
>

mpariente commented:
Great! 
Yes, it would be nice to have it

JunzheJosephZhu commented:
Right, gimme a day or two... I need to buy a new router/SIM card for this
thing...
Joseph Zhu


On Fri, Mar 26, 2021 at 12:28 AM Joseph Zhu ***@***.***>
wrote:

> It seems that I'm unable to connect to my server with ssh, because my new
> apartment has horrible internet connection... Let me try to figure this out.
> Joseph Zhu
>
>
> On Thu, Mar 25, 2021 at 11:56 PM Joseph Zhu ***@***.***>
> wrote:
>
>> Sorry! I was busy with work and forgot. Will do it right away. Btw, I
>> still need to write the bash scipt/inference part. Do I use another commit?
>>
>> On Thu, Mar 25, 2021 at 10:17 PM Pariente Manuel <
>> ***@***.***> wrote:
>>
>>> Let's get this merged? There are very few things to fix :-)
>>>
>>> —
>>> You are receiving this because you authored the thread.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-806833417>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AF2C6GYF6GA3SWRQ46NZNGLTFNAWNANCNFSM4ZCSQUPQ>
>>> .
>>>
>> --
>> Joseph Zhu
>>
>

mpariente commented:
Ping me when this is ready for review

JunzheJosephZhu commented:
I just pushed another commit addressing your comments - I still need to
write the test-time separation method, but should we first get this PR
published?

On Tue, Mar 30, 2021 at 9:00 PM Pariente Manuel ***@***.***>
wrote:

> Ping me when this is ready for review
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-810213121>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF2C6GY5JBMGOE4OKRKVI2DTGHDOZANCNFSM4ZCSQUPQ>
> .
>
-- 
Joseph Zhu

mpariente commented:
Let's merge it with this PR, I'll wait for the separation method to review.

Le mar. 30 mars 2021 à 15:05, Joseph ***@***.***> a écrit :

> I just pushed another commit addressing your comments - I still need to
> write the test-time separation method, but should we first get this PR
> published?
>
> On Tue, Mar 30, 2021 at 9:00 PM Pariente Manuel ***@***.***>
> wrote:
>
> > Ping me when this is ready for review
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/asteroid-team/asteroid/pull/463#issuecomment-810213121
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AF2C6GY5JBMGOE4OKRKVI2DTGHDOZANCNFSM4ZCSQUPQ
> >
> > .
> >
> --
> Joseph Zhu
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/asteroid-team/asteroid/pull/463#issuecomment-810218567>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEND2HCWYSD6G33KKZ7RKLLTGHECTANCNFSM4ZCSQUPQ>
> .
>

JunzheJosephZhu commented:
I just finished writing the eval script. Now only bash script is left. (I don't know how to write bash, btw, so I'd be nice if I could get some help)

JunzheJosephZhu commented:
Also, I'm not entirely sure how to share models on huggingface. Can someone send me a tutorial somewhere? I currently have both .ckpt and .pth

mpariente commented:
You just need to create a repo there, pull, add your model, change the readme and push. 
You can rename the last/best model to `pytorch.bin` and then fill in the README with some tags on the top. Have a look at [this](https://huggingface.co/mpariente/ConvTasNet_Libri1Mix_enhsingle_8k/blob/main/README.md) for example. 

Thanks for your work on this recipe

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/local/conf.yml:
```diff
@@ -0,0 +1,41 @@
+  weight_decay: 0.00000
+# Data config
+data:
+  train_dir: "../../../dataset/{}speakers/wav8k/min/tr"
+  valid_dir: "../../../dataset/{}speakers/wav8k/min/cv"
```
I think I said this before, but relative path in the config file doesn't seem right.

JunzheJosephZhu replied:
ok. I'll change this to data/{}speakers/wav8k/min/tr. I just wrote relative paths because how it was on my system

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/run.sh:
```diff
@@ -0,0 +1,138 @@
+		echo "Generating json files in $tmp_dumpdir"
+		[[ ! -d $tmp_dumpdir ]] && mkdir -p $tmp_dumpdir
+		local_wsj_dir=$wsj0mix_wav_dir/${tmp_nsrc}speakers/wav8k/min/
+		$python_path local/preprocess_wsj0mix.py --in_dir $local_wsj_dir \
```
Does this work?

JunzheJosephZhu replied:
Yeah, this script is just the standard script, but the for loop wraps for n_src

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/run.sh:
```diff
@@ -0,0 +1,138 @@
+
+if [[ $stage -le 3 ]]; then
+  echo "Stage 3: Training"
+  echo "If you want to change n_srcs, please change the config file"
```
Printing this every time doesn't seem useful. 
Also, why n_srcs is in the file if you can't change it.

JunzheJosephZhu replied:
I don't think bash supports adding an integer list as argument, it just gets processed as strings automatically.

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/train.py:
```diff
@@ -0,0 +1,214 @@
+import os
```
clean unused imports

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/train.py:
```diff
@@ -0,0 +1,214 @@
+        self.log("avg_sdr", avg_sdr)
+
+
+class WeightedPITLoss(nn.Module):
```
Can you add a docstring and some comments for understanding what this does on the high level?

mpariente replied:
Is this fixed?

JunzheJosephZhu replied:
Yeah, you can see the docstring above the __init__ line.

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/eval.py:
```diff
@@ -0,0 +1,151 @@
+    [1] "Multi-Decoder DPRNN: High Accuracy Source Counting and Separation",
+        Junzhe Zhu, Raymond Yeh, Mark Hasegawa-Johnson. https://arxiv.org/abs/2011.12022
+"""
+from metrics import Penalized_PIT_Wrapper, pairwise_neg_sisdr_loose
```
Is this anywhere? I don't see it.

mpariente replied:
Is this fixed?

JunzheJosephZhu replied:
yes, I added metrics.py

mpariente reviewed (commented):
Thanks for the PR. Again, some changes before it can be merged.

JunzheJosephZhu commented:
I fixed the stuff you mentioned, could we pass this PR now?

mpariente commented:
Have you managed to share a model on HF Hub?

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/eval.py:
```diff
@@ -0,0 +1,151 @@
+        est_sources = model.separate(mix[None])
+        p_si_snr = Penalized_PIT_Wrapper(pairwise_neg_sisdr_loose)(est_sources, sources)
+        utt_metrics = {
+            "P-Si-SNR": p_si_snr.item(),
```
What's `P-Si-SNR`?

JunzheJosephZhu replied:
It's defined in the paper.

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/eval.py:
```diff
@@ -0,0 +1,151 @@
+        p_si_snr = Penalized_PIT_Wrapper(pairwise_neg_sisdr_loose)(est_sources, sources)
+        utt_metrics = {
+            "P-Si-SNR": p_si_snr.item(),
+            "Accuracy": float(sources.size(0) == est_sources.size(0)),
```
If this is a counting accuracy, then `counting_accuracy` is better.

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/metrics.py:
```diff
@@ -0,0 +1,90 @@
+from scipy.optimize import linear_sum_assignment
+
+
+class PairwiseNegSDR_Loose(_Loss):
```
Loss intead of Loose

mpariente commented on egs/wsj0-mix-var/Multi-Decoder-DPRNN/metrics.py:
```diff
@@ -0,0 +1,90 @@
+        self.take_log = take_log
+        self.EPS = EPS
+
+    def forward(self, est_targets, targets):
```
Is there a way not to duplicate all this code, but subclass it, with minor modif.

JunzheJosephZhu replied:
The main problem is the line assert targets.size() == est_targets.size() in your original code, as it does not support variable number of speakers.

mpariente reviewed (commented):
We are getting close, thanks again for your dedication.

JunzheJosephZhu commented:
Do we have any further issue to fix, or should we pass this PR?

mpariente commented:
We're close.

> Have you managed to share a model on HF Hub?

JunzheJosephZhu commented:
Hi Pariente,
Sorry for the late reply. I added the model to huggingface. 
https://huggingface.co/JunzheJosephZhu/MultiDecoderDPRNN
